proposals/0000-asset-primitive.md

```diff
@@ -0,0 +1,233 @@
+  const allAssets = import.meta.globEager('/assets/**');
+  const path = getPathFromName(name);
+  return allAssets[path];
+} */
+---
+<html>
+  <head>
+    <Font use={Astro.asset('roboto-slab-v22-latin-regular')} family="Roboto Slab" />
+    <style>
+      body { font-family: 'Roboto Slab'; }
+    </style>
+  </head>
+  <body>
+    <p>Hello, <strong>world!</strong></p>
+    <Icon use={Astro.asset('book')} />
+    <Image use={Astro.asset('cookie')} />
+  </body>
+</html>
+```
+
+
+
+# Motivation
+
+We have no official, consistent story around loading assets in Astro. Some plugins have attempted to solve this in userland, for example:
+- Icons: https://github.com/natemoo-re/astro-icon
+- Images: https://github.com/Princesseuh/astro-eleventy-img
+- Fonts: https://github.com/aFuzzyBear/astro-ui/tree/main/components/Fonts
+- Fonts: https://gist.github.com/FredKSchott/3d656427d53a3609ae337c60cc6e3002
+
+All of these approaches work today (Astro v0.21), but suffer from an important problem: they resolve assets at **run-time** (aka "during the render"). For example, a component like `<Icon name="book">` can't know the value for `props.name` until the component and page itself is being run and rendered. This is even better illustrated if the value is a variable, like `<Icon name={fetchedAPIUserData.avatarIcon}>`. Astro would never know what this is.
+
+Why is this is a problem? To be clear, it's not a problem *today* because run-time and build-time are the same. In Astro today, **run-time** and **build-time** both happen during your build, ahead-of-time, before the site is deployed to the server. Runtime has full access to your machine, and every source file in your repo.
+
+However, with SSR support coming this will no longer always be true. If `<Icon name={fetchedAPIUserData.avatarIcon}>` renders for the first time on the server (ex: on Netlify, on Vercel) then Astro has no idea which icon you need to include in your build at build-time, which is when the decision of "what to include in your build" is made.
+
+This RFC is a priority to support our community members who are building components and integrations that use assets, and who need an asset strategy that we can commit to working in the future.  There have been different ideas thrown around to solve for this, some of which I will touch on in the "Alternatives" section of this RFC.
+
+### Goals
+
+This proposal outlines an **asset primitive** that accomplishes the following goals:
+
+- Works for Fonts, Icons, and Images.
+- Can reasonably extend to other asset types in the future.
+- Can reasonably extend to support image optimizations in the future.
+- Works with SSR, where all assets must be known at build-time, with 100% accuracy.
+- Works across frameworks (`.jsx`, `.svelte`, `.vue`). Not an `.astro`-only solution.
+- Introduces an acceptable amount of boilerplate (as little as possible, or none).
+- Can power custom 3rd-party integrations and components.
```

Comment:
How does this power 3rd-party (or even 1st party) integrations? I think that's what's missing most from this proposal.

The only thing from this proposal that really looks different to me from what can be done today is that you can import an SVG without using `?raw`. Is that the only change here, or am I missing something?

Reply:
Clarified to be more clear. The "powering" of 3rd party integration was in reference to the fact that this RFC lets an integration author assume a common data structure for certain assets.

For example, if you are build an image component today, Astro provides no standard idea of what an image is. Today, we leave it to the integration to override and dictate what happens when you import an image in an Astro project. For example: https://github.com/ElMassimo/vite-plugin-image-presets